rustfst-ffi/src/fst.rs

```diff
@@ -0,0 +1,359 @@
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_input_symbols(
+    fst: *const CFst,
+    mut input_symt: *mut CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        fst.input_symbols()
+            .map(|it| {
+                let symt = CSymbolTable(it.clone());
+                unsafe { *input_symt = symt }
+            })
+            .unwrap_or_else(|| input_symt = std::ptr::null_mut());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_input_symbols(
+    fst: *mut CFst,
+    symt: *const CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        let symt = get!(CSymbolTable, symt);
+        fst.set_input_symbols(symt.clone());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_unset_input_symbols(fst: *mut CFst) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.take_input_symbols();
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_output_symbols(
+    fst: *const CFst,
+    mut output_symt: *mut CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        fst.output_symbols()
+            .map(|it| {
+                let symt = CSymbolTable(it.clone());
+                unsafe { *output_symt = symt }
+            })
+            .unwrap_or_else(|| output_symt = std::ptr::null_mut());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_output_symbols(
+    fst: *mut CFst,
+    symt: *const CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        let symt = get!(CSymbolTable, symt);
+        fst.set_output_symbols(symt.clone());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_unset_output_symbols(fst: *mut CFst) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.take_output_symbols();
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_final(
+    fst: *mut CFst,
+    state: CStateId,
+    weight: libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.set_final(state, TropicalWeight::new(weight as f32))?;
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_is_final(
+    fst: *const CFst,
+    state: CStateId,
+    is_final: *mut libc::size_t,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst.is_final(state)?;
+        unsafe { *is_final = res as usize }
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_final_weight(
+    fst: *const CFst,
+    state_id: CStateId,
+    final_weight: *mut libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst
+            .final_weight(state_id)?
+            .ok_or_else(|| anyhow!("State '{:?}' is NOT final", state_id))?;
```

Comment:
Same here, not sure we want to raise an error here. 
We should return None without crashing when there is not final state != raise an error when the state doesn't exist.